From 1da53bdf13d5d1c86e04966a131846c318e4e862 Mon Sep 17 00:00:00 2001
From: "Manjunathappa, Prakash" <prakash.pm@ti.com>
Date: Thu, 16 Feb 2012 21:37:54 +0530
Subject: [PATCH 0789/1092] ARM: OMAP2+: am33xx: Register LCD device via HWMOD
 data

Make necessary changes to register LCDC device via HWMOD data.

Signed-off-by: Manjunathappa, Prakash <prakash.pm@ti.com>
---
 arch/arm/mach-omap2/devices.c | 48 +++++++++++++++++--------------------------
 1 file changed, 19 insertions(+), 29 deletions(-)

diff --git a/arch/arm/mach-omap2/devices.c b/arch/arm/mach-omap2/devices.c
index 56a8bd4..9b5c059 100644
--- a/arch/arm/mach-omap2/devices.c
+++ b/arch/arm/mach-omap2/devices.c
@@ -146,39 +146,29 @@ static struct platform_device omap2cam_device = {
 	.resource	= omap2cam_resources,
 };
 #endif
-#define L4_PER_LCDC_PHYS        0x4830E000
 
-static struct resource am33xx_lcdc_resources[] = {
-	[0] = { /* registers */
-		.start  = L4_PER_LCDC_PHYS,
-		.end    = L4_PER_LCDC_PHYS + SZ_4K - 1,
-		.flags  = IORESOURCE_MEM,
-	},
-	[1] = { /* interrupt */
-		.start  = AM33XX_IRQ_LCD,
-		.end    = AM33XX_IRQ_LCD,
-		.flags  = IORESOURCE_IRQ,
-	},
-};
-
-static struct platform_device am33xx_lcdc_device = {
-	.name		= "da8xx_lcdc",
-	.id		= 0,
-	.num_resources	= ARRAY_SIZE(am33xx_lcdc_resources),
-	.resource	= am33xx_lcdc_resources,
-};
-
-void __init am33xx_register_lcdc(struct da8xx_lcdc_platform_data *pdata)
+int __init am33xx_register_lcdc(struct da8xx_lcdc_platform_data *pdata)
 {
-	int ret;
-
-	am33xx_lcdc_device.dev.platform_data = pdata;
+	int id = 0;
+	struct platform_device *pdev;
+	struct omap_hwmod *oh;
+	char *oh_name = "lcdc";
+	char *dev_name = "da8xx_lcdc";
 
-	ret = platform_device_register(&am33xx_lcdc_device);
-	if (ret)
-		pr_warning("am33xx_register_lcdc: lcdc registration failed: %d\n",
-				ret);
+	oh = omap_hwmod_lookup(oh_name);
+	if (!oh) {
+		pr_err("Could not look up LCD%d hwmod\n", id);
+		return -ENODEV;
+	}
 
+	pdev = omap_device_build(dev_name, id, oh, pdata,
+			sizeof(struct da8xx_lcdc_platform_data), NULL, 0, 0);
+	if (IS_ERR(pdev)) {
+		WARN(1, "Can't build omap_device for %s:%s.\n",
+			dev_name, oh->name);
+		return PTR_ERR(pdev);
+	}
+	return 0;
 }
 
 int __init am33xx_register_tsc(struct tsc_data *pdata)
-- 
1.7.11.2

